src/tool/subcommands/snapshot_cmd.rs

```diff
@@ -366,3 +403,396 @@ fn validation_spinner(prefix: &'static str) -> indicatif::ProgressBar {
+    ///     ("front loaded" GasCharges)   │
+    ///                                  (T)
+    ///
+    /// (T): a CallTree node
+    /// ```
+    ///
+    /// Multiple call trees and trailing gas will be warned and ignored.
+    /// If no call tree is found, returns [`Ok(None)`]
+    fn parse_events(
+        events: Vec<fvm3::trace::ExecutionEvent>,
+    ) -> Result<Option<CallTree>, BuildCallTreeError> {
+        let mut events = VecDeque::from(events);
+        let mut front_load_me = vec![];
+        let mut call_trees = vec![];
+
+        // we don't use a `for` loop so we can pass events them to inner parsers
+        while let Some(event) = events.pop_front() {
+            match event {
+                fvm3::trace::ExecutionEvent::GasCharge(gas_charge) => {
+                    front_load_me.push(gas_charge)
+                }
+                fvm3::trace::ExecutionEvent::Call {
+                    from,
+                    to,
+                    method,
+                    params,
+                    value,
+                } => call_trees.push(CallTree::parse(
+                    ExecutionEventCall {
+                        from,
+                        to,
+                        method,
+                        params,
+                        value,
+                    },
+                    {
+                        // if CallTree::parse took impl Iterator<Item = ExecutionEvent>
+                        // the compiler would infinitely recurse trying to resolve
+                        // &mut &mut &mut ..: Iterator
+                        // so use a VecDeque instead
+                        for gc in front_load_me.drain(..).rev() {
+                            events.push_front(fvm3::trace::ExecutionEvent::GasCharge(gc))
+                        }
+                        &mut events
+                    },
+                )?),
+                fvm3::trace::ExecutionEvent::CallReturn(_, _)
+                | fvm3::trace::ExecutionEvent::CallError(_) => {
+                    return Err(BuildCallTreeError::UnexpectedReturn)
+                }
+                unrecognised => return Err(BuildCallTreeError::UnrecognisedEvent(unrecognised)),
+            }
+        }
+
+        if !front_load_me.is_empty() {
+            // FIXME: https://github.com/ChainSafe/forest/issues/2946
+            // stdout of a `--json` command should be (only) JSON, but we don't
+            // also send `tracing` events there.
+            //
+            // this branch is basically always hit, so keep our stdout json
+            // until logging is fixed up
+            eprintln!(
+                "vm tracing: ignoring {} trailing gas charges",
+                front_load_me.len()
+            );
+        }
+
+        match call_trees.len() {
+            0 => Ok(None),
+            1 => Ok(Some(call_trees.remove(0))),
+            many => {
+                // FIXME: https://github.com/ChainSafe/forest/issues/2946
```

Comment:
#2946 is a tracking issue that includes things like `Log collection story?`. I don't know what this means. Let's refer directly to actionable issues.